### PR TITLE
Add pytest support for unittest.subTests to CI env

### LINKED ISSUE #142157: Some NJT tests are not failing in CI as expected
Example:
```
python test/test_nestedtensor.py -k test_backward_atanh_cuda_float32
```

This test has an xfail from `fill_()` not being implemented, but [it is now](https://github.com/pytorch/pytorch/blob/12b8c2fd8b91fb54abb50b25f3618bd5150125aa/torch/nested/_internal/ops.py#L2363-L2372), so the test should fail as an unexpected success and the xfail should be removed.
https://github.com/pytorch/pytorch/blob/12b8c2fd8b91fb54abb50b25f3618bd5150125aa/test/test_nestedtensor.py#L8175-L8181

I see the expected test failure locally but no signal in CI about this. Another local failure that doesn't show up in CI is:
```
python test/test_nestedtensor.py -k test_compile_forward_nn_functional_softshrink_cpu_float32
```

Again, I expect this test to fail because of an out-of-date xfail specification, and I do see this locally, but not in CI.

Local environment details:
```
Collecting environment information...
PyTorch version: 2.6.0a0+gitcee3f85
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Fedora Linux 38 (Workstation Edition) (x86_64)
GCC version: (GCC) 12.2.0
Clang version: Could not collect
CMake version: version 3.30.5
Libc version: glibc-2.37

Python version: 3.11.5 (main, Sep 11 2023, 13:54:46) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.8.6-100.fc38.x86_64-x86_64-with-glibc2.37
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3070 Ti Laptop GPU
Nvidia driver version: 550.76
cuDNN version: Probably one of the following:
/usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn.so.9.1.0
/usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn_adv.so.9.1.0
/usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn_cnn.so.9.1.0
/usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn_engines_precompiled.so.9.1.0
/usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn_engines_runtime_compiled.so.9.1.0
/usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn_graph.so.9.1.0
/usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn_heuristic.so.9.1.0
/usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn_ops.so.9.1.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        39 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               20
On-line CPU(s) list:                  0-19
Vendor ID:                            GenuineIntel
Model name:                           12th Gen Intel(R) Core(TM) i7-12700H
CPU family:                           6
Model:                                154
Thread(s) per core:                   2
Core(s) per socket:                   14
Socket(s):                            1
Stepping:                             3
CPU(s) scaling MHz:                   16%
CPU max MHz:                          4700.0000
CPU min MHz:                          400.0000
BogoMIPS:                             5376.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l2 cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            544 KiB (14 instances)
L1i cache:                            704 KiB (14 instances)
L2 cache:                             11.5 MiB (8 instances)
L3 cache:                             24 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-19
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Vulnerable: No microcode
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] bert-pytorch==0.0.1a4
[pip3] clip-anytorch==2.6.0
[pip3] CoCa-pytorch==0.1.0
[pip3] dalle2-pytorch==1.14.2
[pip3] ema-pytorch==0.4.8
[pip3] flake8==6.1.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-comprehensions==3.15.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] mypy==1.11.2
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.0
[pip3] nvidia-cublas-cu12==12.1.3.1
[pip3] nvidia-cuda-cupti-cu12==12.1.105
[pip3] nvidia-cuda-nvrtc-cu12==12.1.105
[pip3] nvidia-cuda-runtime-cu12==12.1.105
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.0.2.54
[pip3] nvidia-curand-cu12==10.3.2.106
[pip3] nvidia-cusolver-cu12==11.4.5.107
[pip3] nvidia-cusparse-cu12==12.1.0.106
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.6.77
[pip3] nvidia-nvtx-cu12==12.1.105
[pip3] onnx==1.16.1
[pip3] onnxruntime==1.18.0
[pip3] onnxscript==0.1.0.dev20240621
[pip3] open-clip-torch==2.24.0
[pip3] optree==0.13.0
[pip3] pytorch-triton==3.1.0+cf34004b8a
[pip3] pytorch-warmup==0.1.1
[pip3] rotary-embedding-torch==0.3.3
[pip3] torch==2.6.0a0+git822e8a0
[pip3] torch-fidelity==0.3.0
[pip3] torch_geometric==2.4.0
[pip3] torchao==0.7.0+gitecc53bff
[pip3] torchao==0.7.0+gitecc53bff
[pip3] torchaudio==2.4.0a0+b829e93
[pip3] torchbench==0.1
[pip3] torchlens==0.1.21
[pip3] torchmetrics==1.4.0.post0
[pip3] torchtune==0.4.0.dev20241010+cu121
[pip3] torchvision==0.20.0a0+6279faa
[pip3] torchviz==0.0.2
[pip3] vector-quantize-pytorch==1.14.24
[conda] bert-pytorch              0.0.1a4                   dev_0    <develop>
[conda] clip-anytorch             2.6.0                    pypi_0    pypi
[conda] coca-pytorch              0.1.0                    pypi_0    pypi
[conda] dalle2-pytorch            1.14.2                   pypi_0    pypi
[conda] ema-pytorch               0.4.8                    pypi_0    pypi
[conda] numpy                     1.26.0                   pypi_0    pypi
[conda] nvidia-cublas-cu12        12.1.3.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.1.105                 pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.1.105                 pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.1.105                 pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.1.0.70                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.0.2.54                pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.2.106               pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.4.5.107               pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.1.0.106               pypi_0    pypi
[conda] nvidia-nccl-cu12          2.21.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.6.77                  pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.1.105                 pypi_0    pypi
[conda] open-clip-torch           2.24.0                   pypi_0    pypi
[conda] optree                    0.13.0                   pypi_0    pypi
[conda] pytorch-triton            3.1.0+cf34004b8a          pypi_0    pypi
[conda] pytorch-warmup            0.1.1                    pypi_0    pypi
[conda] rotary-embedding-torch    0.3.3                    pypi_0    pypi
[conda] torch                     2.6.0a0+git822e8a0           dev_0    <develop>
[conda] torch-fidelity            0.3.0                    pypi_0    pypi
[conda] torch-geometric           2.4.0                    pypi_0    pypi
[conda] torchao                   0.7.0+gitecc53bff           dev_0    <develop>
[conda] torchaudio                2.4.0a0+b829e93           dev_0    <develop>
[conda] torchbench                0.1                       dev_0    <develop>
[conda] torchfix                  0.4.0                    pypi_0    pypi
[conda] torchlens                 0.1.21                   pypi_0    pypi
[conda] torchmetrics              1.4.0.post0              pypi_0    pypi
[conda] torchtune                 0.4.0.dev20241010+cu121          pypi_0    pypi
[conda] torchvision               0.20.0a0+6279faa           dev_0    <develop>
[conda] torchviz                  0.0.2                    pypi_0    pypi
[conda] vector-quantize-pytorch   1.14.24                  pypi_0    pypi
```

cc @seemethere @malfet @pytorch/pytorch-dev-infra @mruberry @ZainRizvi @cpuhrsch @bhosmer @drisspg @soulitzer @davidberard98 @YuqingJ

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #142238
* #142243

Fixes #142157

### CODE REVIEW

**Review by malfet — APPROVED** [review]
LGTM, but you can't land as is, as you'll need to tweak xfails at least for NJT tests (good thing it's a stacked PR :) )

**Review by huydhn — APPROVED** [review]

**Comment by jbschlosser** [comment]
@malfet True that, just wanted to make sure the signal shows up before I tweak the xfails.

I may also add a check for the package to exist and hardfail otherwise :)

**Comment by malfet** [comment]
@jbschlosser Please add ci-no-td label, or you might be unpleasantly surprised

File: torch/testing/_internal/opinfo/core.py [hunk 1]
```diff
@@ -1263,6 +1264,16 @@ def test_foo(self, device, dtype, op):
...
 
+        if USE_PYTEST:
+            try:
+                import pytest_subtests  # noqa: F401
+            except ModuleNotFoundError:
+                raise RuntimeError(
+                    "Encountered an OpInfo test with use_subtests=True and pytest-subtests is "
+                    "not installed. The feature will not work correctly within pytest without "
+                    "this package; please install it."
+                ) from None
```
Comment (Skylion007) [inline torch/testing/_internal/opinfo/core.py:1275]: Don't raise from None, raise from the ModuleNotFoundError instance.

File: .ci/pytorch/win-test.sh [hunk 2]
```diff
@@ -35,7 +35,7 @@ if [[ "$BUILD_ENVIRONMENT" == *cuda* ]]; then
 fi
 
 # TODO: Move both of them to Windows AMI
-python -m pip install pytest-rerunfailures==10.3 pytest-cpp==2.3.0 tensorboard==2.13.0
+python -m pip install pytest-rerunfailures==10.3 pytest-cpp==2.3.0 tensorboard==2.13.0 pytest-subtests==0.13.1
```
Comment (huydhn) [inline .ci/pytorch/win-test.sh:38]: Now that you mention Windows, I think you will need to add this to https://github.com/pytorch/pytorch/blob/main/.github/requirements/pip-requirements-macOS.txt for mac

Comment (jbschlosser) [reply .ci/pytorch/win-test.sh:38]: done, thanks!

**Pull request closed without merging**